Conf1 01

zh23jemu opened the pull request:
Draft version of Conf and MergeConf classes

LegenJCdary commented:
nice PR

LegenJCdary commented on src/application.conf:
```diff
@@ -0,0 +1,18 @@
+{
```
Let's stick with json extension for json files. "application.conf.json" would be admissible but personally I don't like it.

LegenJCdary commented on src/main.py:
```diff
@@ -1,17 +1,44 @@
-
-def ingestation():
-    return main()
+# from modules.configs import MergedConf
```
We need a very good reason to merge commented code to main branch. We will not merge commented code to any stable branch.

LegenJCdary commented on src/main.py:
```diff
@@ -1,17 +1,44 @@
+# from modules.misc.arguments import CliInput
+# from modules.misc import utils
+# from modules.outputs import blocks, logging
+import logging
```
Unused import, check pylint test.

LegenJCdary commented on src/main.py:
```diff
@@ -1,17 +1,44 @@
+#     return main(CliInput.parse_arguments(CliInput()))
 
 
 if __name__ == "__main__":
```
Please stick with proposed main.py layout.

LegenJCdary commented on src/main.py:
```diff
@@ -1,17 +1,44 @@
 
 if __name__ == "__main__":
-    ingestation()
+    app_conf = config.Conf("", "application")
```
These are not needed here, merged_conf/final_conf will suffice in main(). configs module and MergedConf class should encapsulate all config functionalities.

LegenJCdary commented on src/modules/configs/templates.py:
```diff
@@ -0,0 +1,108 @@
+    "include": []
+}
+
+conf_v1 = {
```
`merged_conf = {...`

LegenJCdary commented on src/modules/global_vars/global_vars.py:
```diff
@@ -0,0 +1,25 @@
+project_conf = "project.conf"
+operator_conf = "operator.conf"
+
+application_dir = "../../../configs"
```
Should be
```
application_dir = "/etc/ingestation"
application_conf = "/etc/ingestation/ingestation.json"
```

LegenJCdary commented on src/modules/global_vars/global_vars.py:
```diff
@@ -0,0 +1,25 @@
+application_conf = "application.conf"
```
I don't think it's required.

LegenJCdary commented on src/modules/global_vars/global_vars.py:
```diff
@@ -0,0 +1,25 @@
+
+application_dir = "../../../configs"
+
+options = {
```
I don't think it's required.

LegenJCdary commented on src/modules/global_vars/global_vars.py:
```diff
@@ -0,0 +1,25 @@
+    "application_conf": "application.conf"
+}
+
+keys_not_to_be_flatten = [
```
Let's try to pack all below into `config_rules = {}` or something.

LegenJCdary commented on src/modules/misc/arguments.py:
```diff
@@ -0,0 +1,98 @@
+
+class CliInput:
+
+    option_names = {
```
This will probably land in global_vars.py or application.json - just fyi.

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,14 @@
+class TestConf:
+
+    def test_parse_conf(self):
+        app_conf = Conf("../../../src/application.conf", "application")
```
Paths should be `tests/data/...` like in previous comment.

LegenJCdary replied:
Option names in tests don't need to match acrual conf option names e.g.
```
application = {
	opt1: "" // "workdir"
	option2: ""
}

operator = {
	"opt1": ""
}
```

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,14 @@
+from src.modules.configs.config import Conf
+
+
+class TestConf:
```
Some proposals:
- define some mandatory options and check if they are present in the final_conf
- we should definitely check merging/excluding functionality - if final_conf is according to our expectations
- general approach: try to test each function from source code (as long as it makes sense)
- you can use https://github.com/LegenJCdary/ingeSTation/pull/3/files as a starting point

LegenJCdary replied:
"Good" case example - test should `assert True` for Conf.validate_conf()
```
final = {
	"containers": {
		"sata": {
			"device_path": "/dev",
            "mount_path": "/mnt"
		}
	},
	...
	"exclude": ["ABCD1234", "EFGH5678"],
    "include": []
}
```

LegenJCdary replied:
"Bad" cases examples - test should `assert ValidationError` (or whatever json schema returns when failed):
```
final = {
	"containers": {
		"sata_YYY": {
			"device_path": "",
            "mount_path": ""
		}
	},
	...
	"exclude_XXX": ["ABCD1234", "EFGH5678"],
    "include": []
}
```

```
final = {
	"containers": {
		"sata": {
			"device_path": 369, <- int should not be allowed here (json schema)
            "mount_path": ""
		}
	},
	...
	"exclude": [True, None], <- True and None should not be allowed here
    "include": []
}
```

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,166 @@
+
+class Conf:
+
+    def __init__(self, conf_path: str, conf_type: str):
```
means can be str or bool?

zh23jemu replied:
Union[bool, str]

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,166 @@
+                conf.append(key)
+        return conf
+
+    def check_conflicts(self) -> list:
```
yes i found it is not used. so i deleted it

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,166 @@
+    def get_conf_path(self, conf_path: Union[str, bool]) -> str:
+        if not conf_path:
+            try:
+                conf_path = self.get_conf_here()
```
invalid

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,166 @@
+        if not conf_path:
+            try:
+                conf_path = self.get_conf_here()
+                open(conf_path)
```
should close the file

LegenJCdary replied:
Please read and decide: either use os.isfile() or try/except + context manager (`with open...`)
https://therenegadecoder.com/code/how-to-check-if-a-file-exists-in-python/

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,166 @@
+                    for i, item in enumerate(v):
+                        item_key = f"{new_key}{sep}{i}"  # Add index to key
+                        if isinstance(item, dict):
+                            flattened.update(MergedConf.flatten_dict(item, item_key, sep))
```
```
def flatten_dict(self, d, parent_key='', sep='_'):
...
    flattened.update(self.flatten_dict(item, item_key, sep))
```

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,166 @@
+        for key in self.operator_keys:
+            if key not in conf:
+                conf.append(key)
+        return conf
```
conf == src.module.configs.templates.final

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,166 @@
+            return merge_conf
+
+    def overwrite_conf(self, key: str):
+        if key in self.application_keys:
```
- workdir@app e.g. `/var/log/ingestation/${date}.log`
- workdir@proj e.g. `/net/8k3/e0fs01/irods/PLKRA-PROJECTS/MGK-GSR2/9-Upload/ingestation/logs/${date}.log`
- workdir@op e.g. `/net/8k3/e0fs01/users/zjtt99/my_logs/${date}.log`
With this option precedence rule is inverted, so operator > project > app.

LegenJCdary replied:
Will require some kind of `rules = {}` in `application.json` that defines how to approach each option.

zh23jemu commented on tests/modules/configs/test_config.py:
```diff
@@ -1,43 +1,85 @@
-        conf_dict = test_conf.load_conf()
-        assert test_conf.validate_conf(conf_dict) is None
+    def test_parse_conf01(self):
+        app = Conf(os.path.join(test_data_path, "app.json"), "application")
```
Yes, we can just create an object with an empty path. Then validate the dicts from tests/data

LegenJCdary committed:
Move formatting of start timestamp from InitParams to outputs.blocks

Update aspell_exceptions.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,193 @@
+
+    def validate_conf(self, conf: dict) -> None:
+        if self.conf_type == "application":
+            schema_file = application
```
Thinking about something like `schema_file = importlib.import_module(f"schemas.{self.conf_type}")` but that's not very clean (need to read through this first https://stackoverflow.com/questions/477096/python-import-coding-style).

LegenJCdary replied:
For now maybe `schema_file = self.get_schema_file(self.conf_type)` ?

LegenJCdary replied:
Better yet: add schema_file as Conf attribute and provide it later from specific classes e.g.
```
class OperatorConf(Conf):
    """Class for operator configuration files"""

    def __init__(self, conf_path: Union[str, bool]):
        super().__init__(conf_path, "operator", schemas.operator)
```

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,193 @@
+        super().__init__(conf_path, "application")
+
+
+# pylint: disable=too-many-instance-attributes
```
Please increase `max-attributes = 7` in both pylint confs according to your needs.

LegenJCdary replied:
But try to keep it under 10.

LegenJCdary commented on src/global_vars.py:
```diff
@@ -0,0 +1 @@
+APPLICATION_DIR = "/etc/ingestation"
```
Creating global_vars should be mentioned in commit body.

LegenJCdary commented on src/global_vars.py:
```diff
@@ -1 +1,7 @@
+
 APPLICATION_DIR = "/etc/ingestation"
+
+options = {
```
I don't think this dict is required atm, Conf.get_conf_path() is doing precisely that.

LegenJCdary replied:
In the end, we will enable passing project and operator confs as cli arguments, application will never be a user option.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,53 @@
+
+    def get_conf_path(self, conf_path: Union[bool, str]) -> str:
+        if not conf_path:
+            conf_path = self.get_conf_here()
```
After some consideration, I think this is dangerous actually, especially for application (main) conf - user could overwrite that and we have no permissions verification implemented yet. Please remove get_conf_here() for now, maybe we will bring it back later.

LegenJCdary commented on src/main.py:
```diff
@@ -11,6 +12,9 @@ def main(cli_options: dict) -> None:
     blocks.log_starting_messages(logger, init_params, cli_options)
 
+    config = MergedConf(logger, cli_options)
+    print(config.final)
```
We will add nice output via outputs.blocks.Blocks later on. For now, please use `logger.debug`. Since logging module is configured in this project, we should refrain from print() statements as much as possible.

LegenJCdary commented on .gitignore:
```diff
@@ -1,3 +1,14 @@
+
+
+# IntelliJ files
+.idea/
```
I don't mind comments but please sort alphabetically (`sort -o .gitignore .gitignore`) and reduce empty lines (to 1 or 0) - we need to think about users with small screens. ;)

LegenJCdary commented on pyproject.toml:
```diff
@@ -1,2 +1,5 @@
 [build-system]
 requires = ["setuptools", "wheel"]
+
+[tool.coverage.run]
```
Please move it to a new branch/PR (e.g. `git cherry-pick e6569a1414a66745f10e7f766d887f8bd2d7125c` on a new branch). We need to either create some script `tools/run_coverage.(py/sh)` or add it to CI.

LegenJCdary commented on src/final.json:
```diff
@@ -0,0 +1,69 @@
+    "codename": ""
+  },
+  "notify": {
+    "sadmin": "jakub.wierzbowski@aptiv.com"
```
Please change to dummy value.

LegenJCdary commented on src/global_vars.py:
```diff
@@ -0,0 +1,65 @@
+import os
```
Needs to be in modules after all (I'm very sorry!), I am unable to import. Either one is fine:
`src/modules/global_vars.py` -> `from ..global_vars import APPLICATION_DIR, CONFIG_RULES`
`src/modules/global_vars/__init__.py` -> `from modules.global_vars import APPLICATION_DIR, CONFIG_RULES`

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,193 @@
+
+    def flatten_dict(self, dct, parent_key='', sep='_', exclude_keys=None):
+        if exclude_keys is None:
+            exclude_keys = []
```
This can be default directly in kwarg.

zh23jemu replied:
I did that also. PyCharm warns about "Default argument value is mutable". I found this article about it.
[https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/]

LegenJCdary replied:
Very peculiar, illogical even... Thanks for a lesson, goes straight to wiki.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,193 @@
+
+        self.final = self.create_final_conf()
+
+    def flatten_dict(self, dct, parent_key='', sep='_', exclude_keys=None):
```
Most of current code uses ", so let's try to stick with that.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -75,3 +75,119 @@ class ApplicationConf(Conf):
+            if key in self.operator_keys:
+                merge_conf.update(self.operator_keys[key])
+            return merge_conf
+        merge_conf = []
```
Let's try to separate logical blocks/sections with a blank line in longer functions - here I would split before `merge_conf = []` and add a blank line before each return.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -75,3 +75,119 @@ class ApplicationConf(Conf):
+                    flattened[new_key] = value
+        return flattened
+
+    def get_all_conf(self):
```
It's used in 1 place only, so I think we can remove and call directly below.

LegenJCdary replied:
Also `for key in dict:` is supposedly faster than `for x in dict.keys()`. Details in this PEP if you're interested. https://peps.python.org/pep-0234
For sure I haven't read it but I trust the 2nd answer here. ;) https://stackoverflow.com/questions/3294889/iterating-over-dictionaries-using-for-loops.

LegenJCdary replied:
Also, `templates.final` is the template for the config that we want returned form this module. After merging confs, values from `merge_conf()` need to be mapped to `templates.final`.

zh23jemu replied:
So we need a test function for merge_conf() to test if the result is mapped to templates.final?

LegenJCdary commented on src/modules/configs/templates/templates.py:
```diff
@@ -0,0 +1,229 @@
+final = {
```
Templates should be divided into: final, merged, application, project, operator. Vars should be named something generic e.g. `config` so in the module we can do something like
```
temp = templates.final
final = temp.config
```

zh23jemu committed:
Add jsonschema to requirements

zh23jemu committed:
Update .gitignore, aspell_exceptions & pylint

zh23jemu committed:
Add template and schema files

zh23jemu committed:
Add global_vars
- Add APPLICATION_DIR
- Add CONFIG_RULES
- Add MANDATORY_KEYS
- Add PRECEDENCE_RULES

zh23jemu committed:
Create configs.config module with Conf class

zh23jemu committed:
Add ApplicationConf class to configs.config module

zh23jemu committed:
Add ProjectConf class to configs.config module

zh23jemu committed:
Add OperatorConf class to configs.config module

zh23jemu committed:
Create MergedConf class

zh23jemu committed:
Create tests and test data for configs.config module

zh23jemu committed:
Update main.py

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,183 @@
+
+    def create_final_conf(self) -> dict:
+        final_conf = {}
+        for key in templates.merged:
```
This need to be `templates.final`

LegenJCdary commented:
This was just an interim version for discussion purposes, final version merged with https://github.com/LegenJCdary/ingeSTation/pull/23.
